src/main_sm/fork_4/main/full_tracer.cpp

```diff
@@ -825,8 +841,14 @@ zkresult FullTracer::onFinishTx(Context &ctx, const RomCommand &cmd)
         // set refunded gas
         response.gas_refunded = lastOpcodeExecution.gas_refund;
 
-        // Set gas price of last opcode
-        lastOpcodeExecution.gas_cost = lastOpcodeExecution.gas - response.gas_left;
+        //  Set gas price of last opcode if no error and is not a deploy and is not STOP (RETURN + REVERT)
+        if ( (execution_trace.size() > 1) &&
+             (lastOpcodeExecution.op != 0x00 /*STOP opcode*/ ) &&
+             (lastOpcodeExecution.error.size() == 0) &&
+             (response.call_trace.context.to != "0x") )
+        {
+            lastOpcodeExecution.gas_cost = lastOpcodeExecution.gas - fr.toU64(ctx.pols.GAS[*ctx.pStep]);
```

Comment:
I think you also have to set the gas_cost value to the execution trace, as in the js part:
````
lastOpcodeExecution.gas_cost = String(Number(lastOpcodeExecution.gas) - Number(ctx.GAS));
                lastOpcodeCall.gas_cost = lastOpcodeExecution.gas_cost;
````